pkg/status/status.go

```diff
@@ -195,8 +205,85 @@ func (s *status) setStatusCondition(condition configv1.ClusterStatusConditionTyp
 
 // updateStatus makes the API call to update the ClusterOperator
 func (s *status) updateStatus() {
-	_, err := s.configClient.ClusterOperators().UpdateStatus(s.clusterOperator)
+	newClusterOperator, err := s.configClient.ClusterOperators().UpdateStatus(s.clusterOperator)
 	if err != nil {
 		log.Fatalf("Error %v updating ClusterOperator", err)
+	} else {
+		s.clusterOperator = newClusterOperator
 	}
 }
+
+func (s *status) MonitorClusterStatus(stopCh <-chan struct{}) {
+	var (
+		syncs              int
+		successfulSyncs    int
+		hasClusterOperator bool
+	)
+	go wait.Until(func() {
+		// slow poll until we see a cluster operator API, which could be never
+		if !hasClusterOperator {
+			opStatusGV := schema.GroupVersion{
+				Group:   "config.openshift.io",
+				Version: "v1",
+			}
+			err := discovery.ServerSupportsVersion(s.k8sInterface.Discovery(), opStatusGV)
+			if err != nil {
+				log.Infof("ClusterOperator api not present, skipping update (%v)", err)
+				time.Sleep(time.Minute)
+				return
+			}
+			hasClusterOperator = true
+		}
+
+		// create the cluster operator in an initial state if it does not exist
+		if s.clusterOperator == nil {
+			s.setProgressing(fmt.Sprintf("Progressing towards %s", s.version))
+		}
+
+		// Sample the sync channel and see whether we're successfully retiring syncs as a
+		// proxy for "working" (we can't know when we hit level, but we can at least verify
+		// we are seeing some syncs succeeding). Once we observe at least one successful
+		// sync we can begin reporting available and level.
+		select {
+		case err, ok := <-SyncCh:
+			if !ok {
+				// syncCh should only close if the Run() loop exits
+				time.Sleep(5 * time.Second)
+				log.Fatalf("Status sync channel closed but process did not exit in time")
+			}
+			syncs++
+			if err == nil {
+				successfulSyncs++
+			}
+			// grab any other sync events that have accumulated
+			for len(SyncCh) > 0 {
+				if err := <-SyncCh; err == nil {
+					successfulSyncs++
+				}
+				syncs++
+			}
+			// if we haven't yet accumulated enough syncs, wait longer
+			// TODO: replace these magic numbers with a better measure of syncs
+			if successfulSyncs < 3 && syncs < 5 {
```

Comment:
I would appreciate if people would share their opinion on whether or not they believe that we are waiting to process enough syncs.